cmd/clusterctl/client/tree/discovery.go

```diff
@@ -105,25 +121,103 @@ func Discovery(ctx context.Context, c client.Client, namespace, name string, opt
 	controlPlaneMachines := selectControlPlaneMachines(machinesList)
 	for i := range controlPlaneMachines {
 		cp := controlPlaneMachines[i]
-		addMachineFunc(controlPLane, cp)
+		addMachineFunc(controlPlane, cp)
 	}
 
-	if len(machinesList.Items) == len(controlPlaneMachines) {
-		return tree, nil
+	machinePoolList, err := getMachinePoolsInCluster(ctx, c, cluster.Namespace, cluster.Name)
+	if err != nil {
+		return nil, err
 	}
 
 	workers := VirtualObject(cluster.Namespace, "WorkerGroup", "Workers")
-	tree.Add(cluster, workers)
+	// Add WorkerGroup if there are MachineDeployments or MachinePools
+	if len(machinesList.Items) != len(controlPlaneMachines) || len(machinePoolList.Items) > 0 {
+		tree.Add(cluster, workers)
+	}
+
+	if len(machinesList.Items) != len(controlPlaneMachines) { // Add MachineDeployment objects
+		tree.Add(cluster, workers)
+		err = addMachineDeploymentToObjectTree(ctx, c, cluster, workers, machinesList, tree, options, addMachineFunc)
+		if err != nil {
+			return nil, err
+		}
+
+		// Handles orphan machines.
+		if len(machineMap) < len(machinesList.Items) {
+			other := VirtualObject(cluster.Namespace, "OtherGroup", "Other")
+			tree.Add(workers, other)
+
+			for i := range machinesList.Items {
+				m := &machinesList.Items[i]
+				if _, ok := machineMap[m.Name]; ok {
+					continue
+				}
+				addMachineFunc(other, m)
+			}
+		}
+	}
+
+	if len(machinePoolList.Items) > 0 { // Add MachinePool objects
+		tree.Add(cluster, workers)
+		addMachinePoolsToObjectTree(ctx, c, cluster.Namespace, workers, machinePoolList, tree)
+	}
+
+	return tree, nil
+}
 
+func addClusterResourceSetsToObjectTree(ctx context.Context, c client.Client, cluster *clusterv1.Cluster, tree *ObjectTree) {
+	if resourceSetBinding, err := getResourceSetBindingInCluster(ctx, c, cluster.Namespace, cluster.Name); err == nil {
+		resourceSetGroup := VirtualObject(cluster.Namespace, "ClusterResourceSetGroup", "ClusterResourceSets")
+		tree.Add(cluster, resourceSetGroup)
+
+		for _, binding := range resourceSetBinding.Spec.Bindings {
+			resourceSetRefObject := ObjectReferenceObject(&corev1.ObjectReference{
+				Kind:       "ClusterResourceSet",
+				Namespace:  cluster.Namespace,
+				Name:       binding.ClusterResourceSetName,
+				APIVersion: addonsv1.GroupVersion.String(),
+			})
+			tree.Add(resourceSetGroup, resourceSetRefObject, ObjectMetaName("ClusterResourceSet"))
```

Comment:
nit (not blocking)
```suggestion
			tree.Add(resourceSetGroup, resourceSetRefObject))
```
No need to set the meta name to ClusterResourceSet given that it already match the Kind

Reply:
Just ran the linter and removed the CRS ObjectMetaName. Thanks for your help and suggestions on this PR!